python-docker.sh

```diff
@@ -0,0 +1,31 @@
+#!/bin/bash
+# runs given python commands inside a Builder container.
+# just like `.project.py`, the contents of stdout will be read by a YAML parser
+
+set -e
+
+# errcho() { echo "$@" 1>&2; } # writes to stderr, will not be interpreted as YAML
+# errcho "debug"
+
+. .settings.sh
+
+if [ -e .use-python-3.flag ]; then
+    image=py3
+else
+    image=py2
+fi
+
+# skips building image each time
```

Comment:
I had a lot of junk in my builder installation, it was copying hundreds of megs even after I removed a ~2gb file ;) I wasn't sure what your intent/workflow was. You say it should be idempotent but the Docker file contains COPY and RUN commands and the contents of the build directory will change which ensure it gets re-built each time. I get the practicalities of having a container as a sort of an almost-there development environment but definitely prefer immutable images.

Reply:
There are two operations that you are seeing at work I think:

- the build context for Docker being zipped up and sent to the daemon. `.dockerignore` should exclude files from being picked up during this operation, it's missing something that you have there? I remember this being a problem until `.dockerignore` was expanded to include most of my folder's contents.
- the Dockerfile commands: `COPY` should be re-executed if the files involved change, e.g. `requirements.txt`. `RUN` should be re-executed if the command itself changes, or if a previous layer changes of course.

> definitely prefer immutable images.

Me too. We're not there yet but a "virtualenv" image should definitely be immutable enough that in theory people could use builder by pulling it down rather than building it (not that we do that). `settings.yml` is a bit special because it is a configuration (hence should not go in an image) but it's really static as it never changes.

I'll try to get both these and the original PR green.